src/core/net/nat64_translator.cpp

```diff
@@ -0,0 +1,474 @@
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes implementation for the NAT64 translator.
+ *
+ */
+
+#include "nat64_translator.hpp"
+
+#if OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/locator_getters.hpp"
+#include "common/log.hpp"
+#include "net/checksum.hpp"
+#include "net/ip4_types.hpp"
+#include "net/ip6.hpp"
+
+namespace ot {
+namespace Nat64 {
+
+RegisterLogModule("Nat64");
+
+Translator::Translator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+{
+    mNat64Prefix.Clear();
+    mIp4Cidr.Clear();
+}
+
+Message *Translator::NewIp4Message(const Message::Settings &aSettings)
+{
+    Message *message = Get<Ip6::Ip6>().NewMessage(sizeof(Ip6::Header) - sizeof(Ip4::Header), aSettings);
+
+    if (message != nullptr)
+    {
+        message->SetType(Message::kTypeIp4);
+    }
+
+    return message;
+}
+
+Error Translator::SendMessage(Message &aMessage)
+{
+    bool   freed  = false;
+    Error  error  = kErrorDrop;
+    Result result = TranslateToIp6(aMessage);
+
+    VerifyOrExit(result == kForward);
+
+    error = Get<Ip6::Ip6>().SendRaw(aMessage, !OPENTHREAD_CONFIG_IP6_ALLOW_LOOP_BACK_HOST_DATAGRAMS);
+    freed = true;
+
+exit:
+    if (!freed)
+    {
+        aMessage.Free();
+    }
+
+    return error;
+}
+
+Translator::Result Translator::TranslateFromIp6(Message &aMessage)
+{
+    Result          res = kDrop;
+    Ip6::Header     ip6Header;
+    Ip4::Header     ip4Header;
+    AddressMapping *mapping = nullptr;
+
+    // ParseFrom will do basic checks for the message, including the message length and IP protocol version.
+    if (ip6Header.ParseFrom(aMessage) != kErrorNone)
+    {
+        LogWarn("outgoing datagram is not a valid IPv6 datagram, drop");
+        ExitNow(res = kDrop);
+    }
+
+    if (!mNat64Prefix.IsValidNat64() || !ip6Header.GetDestination().MatchesPrefix(mNat64Prefix))
+    {
+        ExitNow(res = kNotTranslated);
+    }
+
+    if (mIp4Cidr.mLength == 0)
+    {
+        // The NAT64 translation is bypassed (will be handled externally)
+        LogDebg("no IPv4 CIDR for NAT64 configured, forward to upper layer");
+        ExitNow(res = kForward);
+    }
+
+    if (ip6Header.GetHopLimit() <= 1)
+    {
+        LogDebg("outgoing datagram hop limit reached, drop");
+        ExitNow(res = kDrop);
+    }
+    ip6Header.SetHopLimit(ip6Header.GetHopLimit() - 1);
```

Comment:
Is decrementing the hop limit specified in an RFC?

Reply:
I'm trying to follow

> Since the translator is a router, as part of forwarding the packet it needs to decrement either the IPv4 TTL (before the translation) or the IPv6 Hop Limit (after the translation).

In RFC6145 Section 4.1

Reply:
Will the host network stack decrement Hop Limit when forwarding between infrastructure interface and Thread?

Reply:
The host network stack will decrease the hop limit.

So seems we don't have to decrease the hop limit here (and no need to check the hop limit above?)